### PR TITLE
chore: Only run "deep tests" in nightly build

### PR DESCRIPTION
**Pull request by robin-aws**
Should address #2010, but will leave that open for a while until we verify it's enough to unblock things.

Initially, this means the large integration test suite is only run for `ubuntu-18.01` on every PR, and "deep tests" refers to running the suite on the current full set of 4 OS runners (`ubuntu-latest`,  `ubuntu-18.01`, `macos-latest` and `windows-latest`). This should work out well since the Ubuntu runners complete the suite in approximately half the time compared to Windows (~15 minutes vs. > 30 minutes), and have four times the concurrent job limit compared to Mac OS.

To back this up with mechanisms, two places in the CI use the GitHub API to check that the deep tests have run successfully:
1. Every PR will check that the last `deep-tests.yml` run on `master` passed. That means if a nightly build fails because of a PR the day before, subsequent PRs will be blocked until the deep suite is fixed (which may just involve rerunning individual jobs if there were transient failures). To merge changes intended to fix those tests themselves, PRs can add the label `skip-deep-tests-check` (since otherwise we're in a catch-22!)
2. The `publish-release.yml` workflow will verify that the last run of `deep-tests.yml` on the tagged commit ran successfully.

Other kinds of expensive and unlikely-to-break tests could be added or moved to this scope in the future. Note that `deep-tests.yml` includes a `workflow_dispatch:` trigger, meaning it can be run manually from here: https://github.com/dafny-lang/dafny/actions/workflows/deep-tests.yml

Also removed several `push` triggers for workflows: no one was paying attention to these builds, and since our branch protection prohibits direct pushes to master these were only happening immediately after merging PRs, and hence were just adding additional CI capacity load for no good reason.

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: .github/workflows/check-for-workflow-run.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
+// Check for a successful previous run of a workflow.
+// This could make a decent reusable action itself.
+module.exports = async ({github, context, core}, workflowID, branch, sha = null) => {
+  // This API doesn't support filtering by SHA, so we just
+  // fetch the first page and scan manually.
+  // That means if the run is fairly old it may be missed,
+  // but that should be rare.
```
Comment (atomb) [inline .github/workflows/check-for-workflow-run.js:7]: I slightly worry that this could be fragile, but in the worst case it would be sufficient to re-trigger the deep tests manually, right?

File: .github/workflows/msbuild.yml [hunk 2]
```diff
@@ -51,8 +49,14 @@ jobs:
     - name: Check whitespace and style
       working-directory: dafny
       run: dotnet tool run dotnet-format -w -s error --check Source/Dafny.sln --exclude Dafny/Scanner.cs --exclude Dafny/Parser.cs
-      
+  
+  check-deep-tests:
+    uses: ./.github/workflows/check-deep-tests-reusable.yml
+    if: "!(contains(github.event.pull_request.labels.*.name, 'skip-deep-tests-check'))"
```
Comment (atomb) [inline .github/workflows/msbuild.yml:55]: Nice addition! Documentation changes, for example, would be nice to not put through the whole battery of tests.

**Review by atomb — APPROVED** [review]

File: .github/workflows/check-for-workflow-run.js (hunk 1 above)
Comment (robin-aws) [reply .github/workflows/check-for-workflow-run.js:7]: Exactly. The default page size for this API is 30 (https://docs.github.com/en/rest/actions/workflow-runs#list-workflow-runs), and if the nightly build hasn't run at all for a month than we have other problems to deal with. :)

File: .github/workflows/msbuild.yml (hunk 2 above)
Comment (robin-aws) [reply .github/workflows/msbuild.yml:55]: Ah, but notice the feature here is to skip checking that the deep tests have run successfully on the nightly build, to deal with the edge case where a previous PR broke only the deep tests and you need another PR to fix them.

I think you're thinking of a label like `skip-integration-tests` that wouldn't run the integration tests at all on a PR. That's a reasonable feature request too and we could follow up with that. I was a little nervous about it because there are some Dafny files hiding in the `docs` folder tree with lit tests that include them.
